src/cluster_legacy.c

```diff
@@ -5955,20 +5958,19 @@ void addShardReplyForClusterShards(client *c, list *nodes) {
         addNodeDetailsToShardReply(c, n);
         clusterFreeNodesSlotsInfo(n);
     }
+    listRelease(nodes);
 }
 
 /* Add to the output buffer of the given client, an array of slot (start, end)
  * pair owned by the shard, also the primary and set of replica(s) along with
  * information about each node. */
 void clusterCommandShards(client *c) {
-    addReplyArrayLen(c, dictSize(server.cluster->shards));
+    addReplyArrayLen(c, server.cluster->nummasters);
```

Comment:
This change becomes much simpler if we replace the dict with a RAX, instead of maintaining this list. It also preserves the structure we built up, which is the Cluster->Shards->nodes mapping.

Reply:
I will explain the rationale to decide to maintain the pointers to masters instead of shards->list-of-nodes.

We initially thought of replacing the Dict with RAX, but that only helps solve one problem. We can maintain sorted shardIDs with RAX, but we still need the corresponding list of <nodes> to be sorted if we want to uphold the shards->nodes mapping; otherwise, the list of <nodes> is useless. We are currently using shards->nodes only for the CLUSTER SHARDS command and not anywhere else. Having RAX, list, and clusterNode overhead to maintain this list just for one command doesn't seem right when it doesn't come with significant performance or memory benefits over an array of pointers to masters. We felt that maintaining the array of pointers to masters is a reasonable tradeoff. Technically, we still honored the shardId->master contract through pointers to masters.

Reply:
> We are currently using shards->nodes only for the CLUSTER SHARDS command and not anywhere else. Having RAX, list, and clusterNode overhead to maintain this list just for one command doesn't seem right when it doesn't come with significant performance or memory benefits over an array of pointers to masters. We felt that maintaining the array of pointers to masters is a reasonable tradeoff. Technically, we still honored the shardId->master contract through pointers to masters.

The thinking was that we would like to use the SHARD ID more in the future as an O(1) lookup, as it's the more logical identifier as opposed to just the Primary ID.